### PR TITLE
Fix Issue 13123 - Disallow throwing contracts for nothrow functions

### PR DESCRIPTION
**Pull request by wolframw**

### CODE REVIEW

File: compiler/src/dmd/semantic3.d [hunk 1]
```diff
@@ -980,10 +981,18 @@ private extern(C++) final class Semantic3Visitor : Visitor
...
+                // BUG: need to disallow returns
                 // BUG: verify that all in and ref parameters are read
                 freq = freq.statementSemantic(sc2);
-                freq.blockExit(funcdecl, false);
+
+                const blockExit = freq.blockExit(funcdecl, false);
+                if (blockExit & BE.throw_)
+                {
+                    if (isnothrow)
+                        error(funcdecl.loc, "`%s`: `in` contract may throw but function is marked as `nothrow`", funcdecl.toPrettyChars());
```
Comment (ibuclaw) [inline compiler/src/dmd/semantic3.d:992]: The usual convention is to  start with `deprecation`, and add an `@@@DEPRECATED_2.1xx@@@` comment - `xx` being the version that it should be made an error, take the next version and add 10.

Comment (wolframw) [reply compiler/src/dmd/semantic3.d]: Thanks, I also added an according changelog entry and opened dlang/dlang.org#3392 to update the list of deprecated features.

**Comment by wolframw** [comment]
Created dlang/phobos#8547 to fix throwing `nothrow` contracts in Phobos.

**Comment by thewilsonator** [comment]
Dependant PR merged.

**Comment by dkorpel** [comment]
@wolframw Can you please `git rebase upstream/master` and force push to restart the test suite with the newest Phobos?

**Review by dkorpel — APPROVED** [review]

**Pull request merged**